lib/PersonalityProvider.jsm

```diff
@@ -0,0 +1,197 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const {PersistentCache} = ChromeUtils.import("resource://activity-stream/lib/PersistentCache.jsm", {});
+const {RemoteSettings} = ChromeUtils.import("resource://services-settings/remote-settings.js", {});
+
+const {NaiveBayesTextTagger} = ChromeUtils.import("resource://activity-stream/lib/NaiveBayesTextTagger.jsm", {});
+const {NmfTextTagger} = ChromeUtils.import("resource://activity-stream/lib/NmfTextTagger.jsm", {});
+const {RecipeExecutor} = ChromeUtils.import("resource://activity-stream/lib/RecipeExecutor.jsm", {});
+
+ChromeUtils.defineModuleGetter(this, "NewTabUtils",
+  "resource://gre/modules/NewTabUtils.jsm");
+
+const STORE_UPDATE_TIME = 24 * 60 * 60 * 1000; // 24 hours
+
+/**
+ * V2 provider builds and ranks an interest profile (also called an “interest vector”) off the browse history.
+ * This allows Firefox to classify pages into topics, by examining the text found on the page.
+ * It does this by looking at the history text content, title, and description.
+ */
+this.PersonalityProvider = class PersonalityProvider {
+  constructor(
+    timeSegments,
+    parameterSets,
+    maxHistoryQueryResults,
+    version,
+    scores,
+    modelKeys) {
+    this.modelKeys = modelKeys;
+    this.timeSegments = timeSegments;
+    this.parameterSets = parameterSets;
+    this.maxHistoryQueryResults = maxHistoryQueryResults;
+    this.version = version;
+    this.scores = scores;
+    this.store = new PersistentCache("personality-provider", true);
+  }
+
+  async init() {
+    this.interestConfig = await this.getRecipe();
+    this.recipeExecutor = await this.generateRecipeExecutor();
+    this.interestVector = await this.store.get("interest-vector");
+
+    // Fetch a new one if none exists or every set update time.
+    if (!this.interestVector ||
+      (Date.now() - this.interestVector.lastUpdate) >= STORE_UPDATE_TIME) {
+      this.interestVector = await this.createInterestVector();
+      this.interestVector.lastUpdate = Date.now();
+      this.store.set("interest-vector", this.interestVector);
+    }
+    this.initialized = true;
+  }
+
+  getRemoteSettings(name) {
+    return RemoteSettings(name).get();
+  }
+
+  getRecipeExecutor(nbTaggers, nmfTaggers) {
+    return new RecipeExecutor(nbTaggers, nmfTaggers);
+  }
+
+  getNaiveBayesTextTagger(model) {
+    return new NaiveBayesTextTagger(model);
+  }
+
+  getNmfTextTagger(model) {
+    return new NmfTextTagger(model);
+  }
+
+  getNewTabUtils() {
```

Comment:
Why wrapping NewTabUtils here? You can simply just do 'await NewTabUtils.activityStreamProvider.executePlacesQuery'

Reply:
I got into the habit of doing this for globals for test purposes because a lot of things (example remote-settings) are globals I cannot set as something else in unit tests. While I only needed to wrap remote settings too make stubbing possible, the rest was just for convenience to make it easier to stub.

Reply:
That's a fair point. `NewtabUtils.activityStreamProvider` already has a default stub defined in `test/unit-entry.js`, is it possible to reuse that here?

Reply:
Yeah, I'm removing all these get functions cept for the needed one, which was renamed. I'm using stubs in the tests instead.